File: src/parser.ts
import { Element } from "./element/element";
import { Expression } from "./element/expression";
import { Type } from "./element/type";
import { Lexer, TokenTypes } from "./lexer";
import { logger } from "./logger";
import { MatchResult } from "./match";
import { Expression as PatternExpression, Literal, Optional, Regex, Sentence, Union } from "./pattern/statement";
import { effects, expressions, structures } from "./registry";
import { Block } from "./statement/block";
import { EffectStatement } from "./statement/effect";
import { ExpressionStatement } from "./statement/expression";
import { StructureStatement } from "./statement/structure";
import { TokenStream } from "./stream";

export class Parser {
    public structures = this.sort(structures);
    public effects = this.sort(effects);
    public expressions = this.sort(expressions);

    constructor(public stream: TokenStream) {}

    public sort<K extends Element, V>(registry: Map<K, V>): [K, V][] {
        return Array.from(registry.entries())
            .sort(([a], [b]) => {
                const priorityA = a.getPriority();
                const priorityB = b.getPriority();

                if (priorityA < priorityB)
                    return 1;
                else if (priorityA === priorityB)
                    return 0;
                else if (priorityA > priorityB)
                    return -1;
            });
    }

    public parse(): StructureStatement[] {
        const structures = [];

        while (!this.stream.isEnd()) {
            const structure = this.parseStructure();

            if (!structure)
                continue;

            structures.push(structure);
        }

        return structures;
    }

    public parseEffect(parent: Block): EffectStatement {
        for (const [effect, pattern] of this.effects) {
            const match = this.matchPattern(pattern.compiledPattern, parent);

            if (!match)
                continue;

            if (!this.stream.isEnd()) {
                if (this.stream.peek().type === TokenTypes.NEWLINE)
                    this.stream.consume().expect(TokenTypes.NEWLINE);
                else continue;
            }

            return new EffectStatement(this, effect, match, parent);
        }

        logger.error("Invalid effect");
        return null;
    }

    public parseStructure(parent?: Block): StructureStatement {
        for (const [structure, pattern] of this.structures) {
            const match = this.matchPattern(pattern.compiledPattern, parent);

            if (!match)
                continue;

            this.stream.peek().expect(TokenTypes.SYMBOL_COLON);
            const block = this.parseBlock();

            if (!this.stream.isEnd())
                this.stream.consume().expect(TokenTypes.NEWLINE);

            return new StructureStatement(this, structure, block, match, parent);
        }

        logger.error("Invalid structure");
        return null;
    }

    public parseBlock(parent?: Block): Block {
        const block = new Block([], parent);
        let indentation = 0;

        this.stream.consume().expect(TokenTypes.SYMBOL_COLON);

        if (!this.stream.isEnd())
            this.stream.consume().expect(TokenTypes.NEWLINE);
        
        if (this.stream.isEnd() || this.stream.peek().type !== TokenTypes.TAB) {
            console.log("[WARN] Empty block");
            return block;
        }

        while (!this.stream.isEnd() && this.stream.peek().type === TokenTypes.TAB) {
            let badEffect = false;

            if (indentation !== 0) {
                for (let i = 0; i < indentation; i++) {
                    if (this.stream.peek().type !== TokenTypes.TAB) {
                        badEffect = true;
                        break;
                    }

                    this.stream.consume().expect(TokenTypes.TAB);
                }

                if (badEffect)
                    break;

                if (this.stream.peek().type === TokenTypes.TAB) {
                    logger.error(`Invalid indentation, expected ${indentation} tabs, found more`);
                    break;
                }
            }

            if (indentation === 0)
                while (this.stream.peek().type === TokenTypes.TAB) {
                    this.stream.consume();
                    indentation++;
                }

            if (indentation === 0) {
                logger.warn("Empty block");
                return block;
            }

            const statement = this.parseEffect(block);

            if (!statement) {
                while (!this.stream.isEnd() && this.stream.peek().type !== TokenTypes.NEWLINE)
                    this.stream.consume();

                if (!this.stream.isEnd())
                    this.stream.consume().expect(TokenTypes.NEWLINE);

                continue;
            }

            block.statements.push(statement);
        }

        return block;
    }

    public parseExpression<T>(type: Type<T>, stream: TokenStream, parent: Block): ExpressionStatement<T> {
        for (const [expression, pattern] of this.expressions) {
            if (expression.getReturnType() !== type.type && type.type !== Object)
                continue;

            const match = this.matchPattern(pattern.compiledPattern, parent, stream);

            if (!match)
                continue;

            return new ExpressionStatement<T>(this, expression as Expression<T>, match, parent);
        }

        logger.error("Invalid expression");
        return null;
    }

    public matchPattern(sentence: Sentence, parent?: Block, origin: TokenStream = this.stream): MatchResult | null {
        const stream = origin.branch();
        const expressions = [];
        const regexes = [];
        const loggerPosition = logger.position.branch();

        for (const index in sentence.data) {
            const first = sentence.data[index];
            const next = sentence.data[parseInt(index) + 1];
            
            if (first instanceof Literal) {
                const token = stream.consume();

                if (first.data != token.value)
                    return null;

                loggerPosition.advance(token.value.length);
            }

            if (first instanceof Optional) {
                if (!this.matchPattern(first.data, parent, stream) && !this.matchPattern(next.data, parent, stream))
                    return null;
            }

            if (first instanceof Union) {
                let matched = false;

                for (const sentence of first.data) {
                    if (this.matchPattern(sentence, parent, stream)) {
                        matched = true;
                        break;
                    }
                }

                if (!matched)
                    return null;
            }

            if (first instanceof Regex) {
                const input = this.constructInput(stream);
                const match = input.match(first.data);

                if (!match)
                    return null;

                regexes.push(match);
                loggerPosition.advance(match[0].length);

                const tokens = new Lexer(match[0])
                    .lex()
                    .filter((t) => t.type !== TokenTypes.WHITESPACE);

                for (const token of tokens)
                    stream.consume().expect(token.type);
            }

            if (first instanceof PatternExpression) {
                const expression = this.parseExpression(first.data.type, stream, parent);

                if (!expression)
                    break;

                expressions.push(expression);
            }
        }

        logger.position.merge(loggerPosition);
        origin.merge(stream);
        return new MatchResult(expressions, regexes);
    }

    public constructInput(origin: TokenStream): string {
        const stream = origin.branch();

        let input = "";

        while (!stream.isEnd()) {
            const first = stream.consume();

            if (first.type === TokenTypes.NEWLINE)
                break;
            
            input += first.value;

            if (first.type === TokenTypes.IDENTIFIER && !stream.isEnd() && stream.peek().type === TokenTypes.IDENTIFIER)
                input += " ";
        }

        return input;
    }
}

File: common/src/main/kotlin/me/honkling/skriptkt/common/parser/Parser.kt
package me.honkling.skriptkt.common.parser

import me.honkling.skriptkt.common.Skript
import me.honkling.skriptkt.common.annotations.ExprType
import me.honkling.skriptkt.common.lexer.ExpectType
import me.honkling.skriptkt.common.lexer.Lexer
import me.honkling.skriptkt.common.lexer.TokenType
import me.honkling.skriptkt.common.node.Block
import me.honkling.skriptkt.common.output.emitError
import me.honkling.skriptkt.common.output.emitWarning
import me.honkling.skriptkt.common.output.emittedErrors
import me.honkling.skriptkt.common.output.emittedWarnings
import me.honkling.skriptkt.common.pattern.element.*
import me.honkling.skriptkt.common.pattern.element.Expression
import me.honkling.skriptkt.common.script.Script
import me.honkling.skriptkt.common.statement.*
import me.honkling.skriptkt.common.syntax.Expression as SyntaxExpression
import me.honkling.skriptkt.common.syntax.*

class Parser(val script: Script, val stream: TokenStream) {
    val file = script.file

    fun parse(): ParseResult {
        val statements = mutableListOf<Statement>()

        while (!stream.isEnd()) {
            val structure = parseStructure() ?: continue

            statements += structure
        }

        return ParseResult(
            file,
            emittedErrors,
            emittedWarnings,
            statements,
            emptyMap())
    }

    private fun parseStructure(): StructureStatement? {
        val structures = Skript.syntaxRegistry.structures
        val result = parseSyntaxElement(structures, null)

        if (result == null) {
            skipBlock()
            emitError("Invalid structure.")
            return null
        }

        val (structure, match) = result
        val block = parseBlock(null)

        if (block == null) {
            emitError("Invalid section. Does the section contain errors?")
            skipBlock()
            return null
        }

        return StructureStatement(this, script, structure.construct() as Structure, block, match)
    }

    private fun parseSection(parent: Block): SectionStatement? {
        val sections = Skript.syntaxRegistry.sections
        val result = parseSyntaxElement(sections, parent)

        if (result == null) {
            emitError("Invalid section.")
            return null
        }

        val (section, match) = result
        val block = parseBlock(null)

        if (block == null) {
            emitError("Invalid section. Does the section contain errors?")
            skipBlock()
            return null
        }

        return SectionStatement(this, parent, section.construct() as Section, block, match)
    }

    private fun parseBlock(parent: Block?): Block? {
        val block = Block(parent, mutableListOf())

        if (stream.consume().expectSymbol(ExpectType.EQUALS, ":") == null)
            return null

        if (stream.isEnd()) {
            emitWarning("Empty section.")
            return block
        }

        if (stream.consume().expect(TokenType.NEWLINE) == null)
            return null

        if (stream.isEnd() || stream.peek().type != TokenType.WHITESPACE) {
            emitWarning("Empty section.")
            return block
        }

        var indentation = -1
        var indentType = null as IndentationType?
        var depth = 1

        while (!stream.isEnd()) {
            if (stream.peek().type == TokenType.NEWLINE)
                stream.consume()

            if (stream.isEnd())
                break

            // Detect indentation of the block & subsequent sections.
            if (indentation == -1) {
                val token = stream.peek()
                indentation = token.raw.length
                indentType = getType(token.raw[0])
            }

            if (stream.peek().type == TokenType.IDENTIFIER)
                break

            val token = stream.consume().expect(TokenType.WHITESPACE)
                ?: return null

            val requirement = indentation * depth
            if (token.raw.length != requirement) {
                emitError("Expected $requirement $indentType, found ${token.raw.length} ${getType(token.raw[0])}")
                return null
            }

            if (!stream.isEnd() && stream.peek().type == TokenType.NEWLINE) {
                stream.consume()
                continue
            }

            val statement = parseEffect(block)
                ?: return null

            if (!stream.isEnd() && stream.consume().expect(TokenType.NEWLINE) == null)
                return null

            block.statements.add(statement)
        }

        return block
    }

    private fun parseEffect(parent: Block): EffectStatement? {
        val effects = Skript.syntaxRegistry.effects
        val result = parseSyntaxElement(effects, parent)

        if (result == null) {
            emitError("Invalid effect.")
            return null
        }

        val (effect, match) = result
        return EffectStatement(this, parent, effect.construct() as Effect, match)
    }

    private fun <T : Any?> parseExpression(type: Type<T>, parent: Block?, stream: TokenStream = this.stream): ExpressionStatement<T>? {
        val expressions = Skript.syntaxRegistry.expressions
        val result = parseSyntaxElement(expressions, parent, stream) {
            val id = it.getAnnotation(ExprType::class.java).type
            id == type.id || id == type.plural
        }

        if (result == null) {
            emitError("Invalid expression.")
            return null
        }

        val (expression, match) = result
        return ExpressionStatement(this, parent, expression.construct() as SyntaxExpression<*>, match) as ExpressionStatement<T>
    }

    fun parseSyntaxElement(
        elements: List<Class<out SyntaxElement>>,
        parent: Block?,
        stream: TokenStream = this.stream,
        filter: (Class<out SyntaxElement>) -> Boolean = { true }
    ): Pair<Class<out SyntaxElement>, PatternMatchResult>? {
        for (element in elements) {
            if (!filter.invoke(element))
                continue

            val pattern = element.getPattern() ?: continue
            return element to (matchPattern(pattern, parent, stream)
                ?: continue)
        }

        return null
    }

    private fun matchPattern(pattern: Sentence, parent: Block?, stream: TokenStream = this.stream): PatternMatchResult? {
        val expressions = mutableListOf<ExpressionStatement<*>>()
        val regexes = mutableListOf<MatchResult>()
        val tags = mutableListOf<String>()
        val branch = stream.branch()

        top@for (patternElement in pattern.elements) {
            when (patternElement) {
                is Literal -> {
                    val token = branch.peek()
                    val raw = token.raw.lowercase()
                    val value = patternElement.value.lowercase()

                    if (raw != value) {
                        if (raw.startsWith(value)) {
                            token.raw = token.raw.substring(value.length)
                            token.start += value.length
                            continue
                        }

                        return null
                    }

                    branch.consume()
                }

                is Optional -> {
                    val match = matchPattern(patternElement.sentence, parent, branch) ?: continue
                    regexes += match.regexes
                    expressions += match.expressions
                    tags += match.tags
                }

                is Choice -> {
                    for ((choice, tag) in patternElement.choices) {
                        val match = matchPattern(choice, parent, branch) ?: continue

                        regexes += match.regexes
                        expressions += match.expressions
                        if (tag != null)
                            tags += tag

                        continue@top
                    }

                    return null
                }

                is RegEx -> {
                    val input = constructInput(branch)
                    val match = patternElement.regex.find(input)
                        ?: return null

                    // todo: is there a better way than lexing the match?
                    for (token in Lexer(match.value).lex()) {
                        val currentToken = branch.peek()
                        val currentRaw = currentToken.raw.lowercase()
                        val raw = token.raw.lowercase()

                        if (currentRaw != raw && currentRaw.startsWith(raw)) {
                            currentToken.raw = currentToken.raw.substring(raw.length)
                            currentToken.start += raw.length
                            continue
                        }

                        branch.consume().expect(token.type)
                    }

                    regexes += match
                }

                is Expression<*> -> {
                    val type = patternElement.type
                    val expression = parseExpression(type, parent, branch)
                        ?: return null

                    expressions.add(expression)
                }
            }

            skipWhitespace(branch)
        }

        stream.merge(branch)
        return PatternMatchResult(regexes, expressions, tags)
    }

    private fun constructInput(stream: TokenStream): String {
        val builder = StringBuilder()
        val branch = stream.branch()

        while (!branch.isEnd()) {
            val first = branch.consume()
            val next = branch.peekOrLastToken()

            if (first.type == TokenType.NEWLINE)
                break

            builder.append(first.raw)

            if (!branch.isEnd() && first.type == TokenType.IDENTIFIER && next.type == TokenType.IDENTIFIER)
                builder.append(" ")
        }

        return builder.toString()
    }

    private fun skipBlock() {
        while (!stream.isEnd()) {
            val token = stream.consume()

            if (!stream.isEnd() && token.type == TokenType.NEWLINE && stream.peek().type == TokenType.IDENTIFIER)
                break
        }
    }

    private fun skipWhitespace(stream: TokenStream) {
        if (!stream.isEnd() && stream.peek().type == TokenType.WHITESPACE)
            stream.consume()
    }
}